Creates a separate crate for NBT encoding/decoding.

atheriel opened the pull request:
This PR supersedes #39.

A few people expressed interest in doing this, and it seemed like a nice way to compartmentalize a piece of the project others might find useful. I've called the crate "hem-nbt" for now, but suggestions are welcome.

Note that the crate is entirely local. It uses the `path` mechanism of cargo in the top-level package.

Thoughts?

atheriel committed:
nbt: Adds the cannonical `bigtest.nbt` test and a few benchmarks.

Results:
    nbt_bench_bigwrite   ... bench:     11154 ns/iter (+/- 6415)
    nbt_bench_smallwrite ... bench:       788 ns/iter (+/- 93)

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
nbt: There is a now a separate library for NBT encoding/decoding.

This commit moves the nbt module into its own local cargo project.
There are some ugly side effects for now: `ReadExactExt` is pasted in
the new file, and the `Protocol` implementation for `NbtBlob` is now in
the `types/mod.rs` file.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

fenhl commented:
Since we have the `hematite_server` precedent, I suggest `hematite_nbt` for the crate name. I would also consider adding the crate to [crates.io](https://crates.io/).

fenhl commented:
Or maybe just `nbt`, which [is still available](https://crates.io/crates/nbt).

atheriel commented:
I felt a little uncomfortable claiming the `nbt` namespace, but maybe that's silly. Also, I believe it's too soon to release it to crates.io -- I'd want to add a number of features (similar to [this Python package](https://github.com/twoolie/NBT) before doing so).

fenhl commented on src/types/mod.rs:
```diff
@@ -1,16 +1,34 @@
 pub use self::slot::Slot;
 pub use self::varnum::Var;
+
+impl Protocol for NbtBlob {
```
This should probably be in a submodule.

atheriel replied:
Yes. I could revive the `nbt` submodule just for this entry.

fenhl commented:
Which features do you want to add? I've worked with the Python package, but it's not very well documented and what I've used it for can already be done with this crate.

atheriel commented:
- Iteration over (name, tag) pairs.
- Pretty printing.
- Better ReadExactExt.
- Also, bug fixes.

fenhl commented:
Sounds good. We can wait with crates.io, it seems like we're the only project doing this right now anyway.

What do you think about my crate name suggestions?

atheriel commented:
Should we be a little audacious and just claim `nbt`?

bvssvni commented:
It is possible to prefix the package name with `hematite-` while using `nbt` as library name.

I'm OK with claiming `nbt` on crates.io, because it's not a generic name that I suspect will be used for something else.

toqueteos commented:
If `nbt` on crates.io is free we are the best candidate out there opting for it, after all it's a Minecraft only format and we are using it for Minecraft purposes so... :+1: :tada:

atheriel committed:
nbt: Revives the `types::nbt` file for a `Protocol` implementation.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel committed:
nbt: Package/library rename. Also fixes #41.

There has been some re-organizing, and the introduction of a new kind
of error to deal with incomplete or incorrect reads.

Signed-off-by: Aaron Jacobs <atheriel@gmail.com>

atheriel commented:
Finally had time to get back to this. I'm holding off pushing it to crates.io for now, and I've taken @bvssvni 's advice about naming in the meantime. Rebased onto master as well.

@toqueteos @fenhl If there are any outstanding issues you see, let me know.
